### PR TITLE
[CUDA] Optimize sumproduct_pair for certain inputs

### PR DESCRIPTION
**Pull request by heitorschueroff**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48184 [CUDA] Optimize sumproduct_pair for certain inputs**
* #47860 Fixed einsum compatibility/performance issues (#46398)

The following benchmark demonstrates that when the batch dimensions are the same size (as opposed to one of them being size 1) and batch sizes < 1024 or and the contraction dimensions add to at most size 256 elements, ``mul`` and ``sum`` is faster than calling ``bmm`` on CUDA.

The script below compares mul/sum against the before/after einsum.

## Benchmark script
```python
import torch
import pickle
import numpy as np
from torch.utils import benchmark

results = []
for b in np.power(2, range(0, 21, 2)):
    for n in np.power(2, range(0, 21, 2)):
        if b * n >= 1000000000:
            continue

        description = f'[{b}, {n}]'
        x = torch.rand(b, n, device='cuda')
        y = torch.rand(b, n, device='cuda')

        results.append(benchmark.Timer(
            stmt='(x * y).sum(dim=-1)',
            globals={'x': x, 'y': y},
            description=description,
            # num_threads=8,
        ).blocked_autorange())

        results.append(benchmark.Timer(
            stmt='torch.einsum("...j,...j->...", x, y)',
            sub_label='New Einsum',
            globals={'x': x, 'y': y},
            description=description,
            # num_threads=8,
        ).blocked_autorange())

# pickle.dump(results, open('bench.p', 'wb'))
results.extend(pickle.load(open('bench.p', 'rb')))

compare = benchmark.Compare(results)
compare.trim_significant_figures()
compare.colorize()
compare.print()
```

## Benchmark results
```
[--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------  --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------]
                                            |  [1, 1]  |  [1, 4]  |  [1, 16]  |  [1, 64]  |  [1, 256]  |  [1, 1024]  |  [1, 4096]  |  [1, 16384]  |  [1, 65536]  |  [1, 262144]  |  [1, 1048576]  |  [4, 1]  |  [4, 4]  |  [4, 16]  |  [4, 64]  |  [4, 256]  |  [4, 1024]  |  [4, 4096]  |  [4, 16384]  |  [4, 65536]  |  [4, 262144]  |  [4, 1048576]  |  [16, 1]  |  [16, 4]  |  [16, 16]  |  [16, 64]  |  [16, 256]  |  [16, 1024]  |  [16, 4096]  |  [16, 16384]  |  [16, 65536]  |  [16, 262144]  |  [16, 1048576]  |  [64, 1]  |  [64, 4]  |  [64, 16]  |  [64, 64]  |  [64, 256]  |  [64, 1024]  |  [64, 4096]  |  [64, 16384]  |  [64, 65536]  |  [64, 262144]  |  [64, 1048576]  |  [256, 1]  |  [256, 4]  |  [256, 16]  |  [256, 64]  |  [256, 256]  |  [256, 1024]  |  [256, 4096]  |  [256, 16384]  |  [256, 65536]  |  [256, 262144]  |  [256, 1048576]  |  [1024, 1]  |  [1024, 4]  |  [1024, 16]  |  [1024, 64]  |  [1024, 256]  |  [1024, 1024]  |  [1024, 4096]  |  [1024, 16384]  |  [1024, 65536]  |  [1024, 262144]  |  [1024, 1048576]  |  [4096, 1]  |  [4096, 4]  |  [4096, 16]  |  [4096, 64]  |  [4096, 256]  |  [4096, 1024]  |  [4096, 4096]  |  [4096, 16384]  |  [4096, 65536]  |  [4096, 262144]  |  [16384, 1]  |  [16384, 4]  |  [16384, 16]  |  [16384, 64]  |  [16384, 256]  |  [16384, 1024]  |  [16384, 4096]  |  [16384, 16384]  |  [16384, 65536]  |  [65536, 1]  |  [65536, 4]  |  [65536, 16]  |  [65536, 64]  |  [65536, 256]  |  [65536, 1024]  |  [65536, 4096]  |  [65536, 16384]  |  [262144, 1]  |  [262144, 4]  |  [262144, 16]  |  [262144, 64]  |  [262144, 256]  |  [262144, 1024]  |  [262144, 4096]  |  [1048576, 1]  |  [1048576, 4]  |  [1048576, 16]  |  [1048576, 64]  |  [1048576, 256]  |  [1048576, 1024]
1 threads: ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
      New Einsum                            |    31    |    40    |     40    |     40    |     40     |      41     |      41     |      40      |      40      |      62.0     |      51.8      |    41    |    44    |     55    |    38.9   |    38.9    |     38.9    |      39     |      39      |      39      |        46     |       140      |     41    |     40    |     40     |     40     |      40     |      39      |      39      |       39      |      44.8     |       144      |        500      |     30    |     40    |     40     |     40     |      40     |      40      |      40      |       41      |      128      |       506      |       1940      |     31     |     40     |      41     |      42     |      42      |       41      |      44.1     |      130       |     488.7      |       1956      |       7790       |      30     |      40     |      40      |      41      |       61      |      39.8      |       89       |      327.7      |       1305      |       5218       |       21400       |      30     |      40     |      40      |      40      |      39.7     |       91       |     337.4      |       1285      |       5152      |      20660       |      29      |      40      |       39      |      50.5     |      106       |      347.0      |       1320      |       4906       |      20000       |      29      |      40      |      70.8     |      192      |     410.0      |       1368      |       5281      |      20000       |       30      |        68     |       274      |       759      |       1661      |       5383       |      21750       |      29.4      |       263      |       1089      |       3034      |       6700       |       21650     
      (x * y).sum(dim=-1)                   |    22    |    21    |     21    |     21    |     22     |      22     |      22     |      22      |      22      |      27.1     |      51.6      |    22    |    22    |     22    |    33.0   |    32.9    |     33.4    |      21     |      22      |      22      |        52     |       145      |     21    |     21    |     21     |     22     |      21     |      21      |      22      |       21      |      44.8     |       144      |        501      |     21    |     21    |     21     |     22     |      21     |      21      |      21      |       36      |      128      |       508      |       1940      |     21     |     22     |      22     |      22     |      22      |       21      |      44.2     |      131       |     490.6      |       1950      |       7790       |      22     |      21     |      21      |      21      |       21      |      34.7      |      128       |      486.8      |       1905      |       7580       |       30400       |      21     |      21     |      22      |      21      |      36.0     |      129       |     495.1      |       1911      |       7580      |      30370       |      21      |      22      |       23      |      50.6     |      129       |      484.2      |       1910      |       7590       |      30270       |      22      |      22      |      70.9     |      192      |     486.3      |       1911      |       7600      |      30270       |       21      |        69     |       274      |       759      |       1921      |       7600       |      30300       |      51.4      |       263      |       1088      |       3036      |       7646       |       30340     
      torch.einsum("...j,...j->...", x, y)  |    28    |    60    |     39    |     39    |     39     |      39     |      40     |      50      |      50      |      49.7     |      49.6      |    28    |    38    |     52    |    53.1   |    53.5    |     52.8    |      37     |      68      |     268      |      1610     |      6452      |     27    |     37    |     37     |     38     |      37     |      38      |      38      |       68      |     409.4     |      1635      |       6530      |     27    |     37    |     37     |     37     |      37     |      37      |      37      |      112      |      463      |      1800      |       7200      |     27     |     38     |      37     |      37     |      37      |       38      |      37.6     |      129       |     511.5      |       2029      |       8080       |      27     |      37     |      37      |      37      |       38      |      37.6      |       88       |      327.4      |       1300      |       5208       |       21360       |      28     |      37     |      38      |      38      |      37.4     |       92       |     337.6      |       1286      |       5165      |      20650       |      27      |      62      |       62      |      74.1     |      107       |      347.3      |       1323      |       4919       |      19670       |      27      |     242      |     276.0     |      286      |     411.0      |       1364      |       5253      |      19600       |       28      |      1082     |      1098      |      1140      |       1689      |       5435       |      21710       |      27.3      |      4317      |       4389      |       4549      |       6660       |       21580     

Times are in microseconds (us).
```

The following benchmark is on CPU. There are some cases where mul/sum is faster then bmm but the pattern here is less clear to know when to chose one over the other.

```
[--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------  -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------]
                           |  [1, 1]  |  [1, 4]  |  [1, 16]  |  [1, 64]  |  [1, 256]  |  [1, 1024]  |  [1, 4096]  |  [1, 16384]  |  [1, 65536]  |  [1, 262144]  |  [1, 1048576]  |  [4, 1]  |  [4, 4]  |  [4, 16]  |  [4, 64]  |  [4, 256]  |  [4, 1024]  |  [4, 4096]  |  [4, 16384]  |  [4, 65536]  |  [4, 262144]  |  [4, 1048576]  |  [16, 1]  |  [16, 4]  |  [16, 16]  |  [16, 64]  |  [16, 256]  |  [16, 1024]  |  [16, 4096]  |  [16, 16384]  |  [16, 65536]  |  [16, 262144]  |  [16, 1048576]  |  [64, 1]  |  [64, 4]  |  [64, 16]  |  [64, 64]  |  [64, 256]  |  [64, 1024]  |  [64, 4096]  |  [64, 16384]  |  [64, 65536]  |  [64, 262144]  |  [64, 1048576]  |  [256, 1]  |  [256, 4]  |  [256, 16]  |  [256, 64]  |  [256, 256]  |  [256, 1024]  |  [256, 4096]  |  [256, 16384]  |  [256, 65536]  |  [256, 262144]  |  [256, 1048576]  |  [1024, 1]  |  [1024, 4]  |  [1024, 16]  |  [1024, 64]  |  [1024, 256]  |  [1024, 1024]  |  [1024, 4096]  |  [1024, 16384]  |  [1024, 65536]  |  [1024, 262144]  |  [4096, 1]  |  [4096, 4]  |  [4096, 16]  |  [4096, 64]  |  [4096, 256]  |  [4096, 1024]  |  [4096, 4096]  |  [4096, 16384]  |  [4096, 65536]  |  [16384, 1]  |  [16384, 4]  |  [16384, 16]  |  [16384, 64]  |  [16384, 256]  |  [16384, 1024]  |  [16384, 4096]  |  [16384, 16384]  |  [65536, 1]  |  [65536, 4]  |  [65536, 16]  |  [65536, 64]  |  [65536, 256]  |  [65536, 1024]  |  [65536, 4096]  |  [262144, 1]  |  [262144, 4]  |  [262144, 16]  |  [262144, 64]  |  [262144, 256]  |  [262144, 1024]  |  [1048576, 1]  |  [1048576, 4]  |  [1048576, 16]  |  [1048576, 64]  |  [1048576, 256]
8 threads: --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
      (x * y).sum(dim=-1)  |     6    |    6.3   |     10    |     10    |     10     |      6.8    |      8.1    |      16      |      63      |      55.3     |      116       |     6    |    6.1   |     6.3   |     6.5   |     7.0    |      8.5    |      16     |     49.5     |       80     |       239     |       890      |      6    |     6.5   |     6.4    |     7.0    |      8.4    |      15      |       48     |        42     |        99     |       317      |       11700     |      6    |     7.3   |     7.4    |     9.4    |      16     |     47.1     |       44     |       100     |       321     |      12000     |       50960     |      6     |     10     |      10     |      18     |     49.6     |       45      |      100      |       321      |     10000      |      51560      |      201300      |      10     |     30.0    |     31.5     |     50.3     |       44      |      100       |       333      |      10000      |      47000      |      200000      |      10     |     81.2    |     52.8     |      48      |      110      |       336      |     10000      |      46880      |      207300     |     37.7     |     92.9     |       63      |      120      |     334.8      |      11300      |      52790      |      200000      |      80      |     230      |      190      |     401.1     |     13000      |      51260      |      212200     |       80      |      710      |     580.0      |     12000      |      52700      |      215100      |      228       |      2783      |      12800      |      51880      |      197000    
      New Einsum           |    14    |   31.8   |     31    |     24    |     24     |     25.0    |     26.6    |      32      |      39      |      49.8     |       74       |    14    |   29.8   |    30.3   |    30.1   |    30.5    |     67.0    |     156     |    610.8     |     2376     |      8600     |     34820      |     13    |    29.8   |    30.0    |    30.5    |     31.5    |     283      |     1000     |      4100     |     20000     |     57000      |      278900     |     13    |    29.8   |    34.7    |    32.3    |      34     |    425.8     |     2000     |      6000     |     23330     |     100000     |      357000     |     14     |     31     |      32     |      33     |     41.1     |      160      |      506      |      2000      |      8000      |      31640      |      136000      |      21     |     42.2    |     34.9     |     40.7     |       68      |      498       |      1800      |       7500      |      30480      |      119000      |      17     |     36.1    |     43.6     |      72      |      170      |      1860      |     10000      |      29820      |      120000     |     34.2     |     54.5     |       80      |      183      |     584.5      |       8000      |      30000      |      120000      |      46      |     100      |      226      |     624.8     |      2971      |      33760      |      200000     |       41      |      362      |     712.9      |      2788      |      20000      |      100000      |       84       |      1000      |       3214      |      17000      |       52650    

Times are in microseconds (us).
```

### TODO

- [ ] Compare results on Volta GPU for half size
- [ ] Verify changes do not impact _trilinear

Differential Revision: [D25059362](https://our.internmc.facebook.com/intern/diff/D25059362)

### CODE REVIEW

**Comment by ngimel** [comment]
What is the difference between the first table and the second table?

**Comment by mruberry** [comment]
More before/after comparisons with different cases, NO discontiguous inputs (uninteresting for performance), + sanity checking the performance surface of the change, since it affects operators other than einsum.

@ngimel: not expanding batch dimensions seems not very interesting

File: aten/src/ATen/native/Linear.cpp [hunk 1]
```diff
@@ -76,6 +76,16 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
       ro_size *= right.size(i);
     }
   }
+
+  if (left_.is_cuda()) {
+    // Experiments showed that for the cases below it is faster to multiply and sum
```
Comment (mruberry) [inline aten/src/ATen/native/Linear.cpp:81]: This is a great comment.

**Comment by heitorschueroff** [comment]
Closing this PR as it introduces regressions for bilinear which also depends indirectly on sumproduct_pair. This probably needs a redesign to break down sumproduct_pair into smaller functions.

**Pull request closed without merging**